src/job-cache/job-cache.cpp

```diff
@@ -0,0 +1,1033 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Open Group Base Specifications Issue 7
+#define _XOPEN_SOURCE 700
+#define _POSIX_C_SOURCE 200809L
+
+#include "job-cache.h"
+
+#include "unique_fd.h"
+
+namespace {
+
+using namespace job_cache;
+
+// moves the file or directory, crashes on error
+static void rename_no_fail(const char *old_path, const char *new_path) {
+  if (rename(old_path, new_path) < 0) {
+    log_fatal("rename(%s, %s): %s", old_path, new_path, strerror(errno));
+  }
+}
+
+// Ensures the the given directory has been created
+static void mkdir_no_fail(const char *dir) {
+  if (mkdir(dir, 0777) < 0 && errno != EEXIST) {
+    log_fatal("mkdir(%s): %s", dir, strerror(errno));
+  }
+}
+
+// Ensures the given file has been deleted
+static void unlink_no_fail(const char *file) {
+  if (unlink(file) < 0) {
+    log_fatal("unlink(%s): %s", file, strerror(errno));
+  }
+}
+
+// Ensures the the given directory no longer exists
+static void rmdir_no_fail(const char *dir) {
+  if (rmdir(dir) < 0 && errno != ENOENT) {
+    log_fatal("rmdir(%s): %s", dir, strerror(errno));
+  }
+}
+
+// This function first attempts to reflink but if that isn't
+// supported by the filesystem it copies instead.
+#ifdef __APPLE__
+
+static void copy(int src_fd, int dst_fd) {
+  // TODO: Actully make this work. APFS supports reflinking so
+  //       it should be possible to do this correctly
+}
+
+#elif __GLIBC__ >= 2 && __GLIBC_MINOR__ >= 27
+
+#include <linux/fs.h>
+
+// This function just uses `copy_file_range` to make
+// an efficent copy
+static void copy(int src_fd, int dst_fd) {
+  struct stat buf;
+  // There's a race here between the fstat and the copy_file_range
```

Comment:
I don't actually remember what the heck I meant here. I think what I meant was that the inode could be modified between the fstat and the copy_file_range operation and that would cause an error. I think we're safe assuming that no one is going to modify these functions though. I was just being hyper aware of the atomicity of all the file operations here for obvious reasons. I'm not sure of a way to solve this to make `copy_file_range` atomic.